File: src/core/canonicalize.ts
/**
 * RFC 8785 — JSON Canonicalization Scheme (JCS).
 *
 * This is the function that decides whether two independent implementations can
 * agree on what a figure IS. If TypeScript and Python disagree on one byte here,
 * every digest, every artifact identity, and every reproducibility claim in the
 * project is worthless. So it is implemented deliberately, tested against the
 * official RFC 8785 vectors, and never described as "sorted JSON.stringify" —
 * that is a different thing that happens to look similar.
 *
 * The scheme, exactly:
 *
 *   - Object members are sorted by their names, compared as sequences of UTF-16
 *     code units (RFC 8785 §3.2.3). JavaScript's default string `<` and
 *     `Array.prototype.sort()` already compare UTF-16 code units, which is why
 *     a bare `.sort()` is correct here and a locale-aware collator would not be.
 *   - Numbers use the ECMAScript Number-to-String algorithm (§3.2.2.3), which is
 *     what `JSON.stringify` emits. `-0` serializes as `0`.
 *   - Strings use the shortest legal JSON escapes (§3.2.2.2) — which is what
 *     `JSON.stringify` emits.
 *   - No insignificant whitespace anywhere.
 *
 * The JCS domain is finite, well-formed JSON. Values outside it — NaN, Infinity,
 * a lone surrogate — are REJECTED rather than coerced, because there is no
 * canonical form for a value the scheme does not define.
 */

import { sha256Digest } from './sha256.js';

export class CanonicalizationError extends Error {
  readonly path: string;
  constructor(message: string, path: string) {
    super(message);
    this.name = 'CanonicalizationError';
    this.path = path;
  }
}

/** A value that is inside the JCS domain. */
export type JsonValue =
  | null
  | boolean
  | number
  | string
  | readonly JsonValue[]
  | { readonly [key: string]: JsonValue };

/**
 * A lone surrogate has no UTF-8 encoding, so it has no canonical byte sequence
 * and cannot be hashed reproducibly. It is a defect at the boundary, not
 * something to paper over with a replacement character.
 */
function assertWellFormed(text: string, path: string): void {
  for (let i = 0; i < text.length; i++) {
    const code = text.charCodeAt(i);
    if (code >= 0xd800 && code <= 0xdbff) {
      const next = i + 1 < text.length ? text.charCodeAt(i + 1) : 0;
      if (!(next >= 0xdc00 && next <= 0xdfff)) {
        throw new CanonicalizationError('unpaired high surrogate', path);
      }
      i++;
    } else if (code >= 0xdc00 && code <= 0xdfff) {
      throw new CanonicalizationError('unpaired low surrogate', path);
    }
  }
}

function serializeNumber(value: number, path: string): string {
  if (!Number.isFinite(value)) {
    throw new CanonicalizationError(
      'non-finite numbers are outside the JCS domain and have no canonical form',
      path,
    );
  }
  // JSON.stringify implements ECMAScript Number-to-String, and maps -0 to "0".
  return JSON.stringify(value);
}

function safeOwnKeys(value: object, path: string): (string | symbol)[] {
  try {
    return Reflect.ownKeys(value);
  } catch {
    throw new CanonicalizationError('object keys could not be inspected without executing a hostile trap', path);
  }
}

function safeDescriptor(value: object, key: PropertyKey, path: string): PropertyDescriptor {
  try {
    const descriptor = Object.getOwnPropertyDescriptor(value, key);
    if (descriptor) return descriptor;
  } catch {
    // Reframed below as a deterministic domain error.
  }
  throw new CanonicalizationError('an object member could not be inspected safely', path);
}

function childPath(path: string, key: string): string {
  return `${path}/${key.replace(/~/gu, '~0').replace(/\//gu, '~1')}`;
}

function serialize(value: unknown, path: string, depth: number): string {
  if (depth > 128) {
    throw new CanonicalizationError('value nests deeper than the canonicalizer permits', path);
  }

  if (value === null) return 'null';

  switch (typeof value) {
    case 'boolean':
      return value ? 'true' : 'false';

    case 'number':
      return serializeNumber(value, path);

    case 'string':
      assertWellFormed(value, path);
      return JSON.stringify(value);

    case 'object':
      break;

    default:
      throw new CanonicalizationError(
        `values of type ${typeof value} are outside the JCS domain`,
        path,
      );
  }

  let array = false;
  try {
    array = Array.isArray(value);
  } catch {
    throw new CanonicalizationError('the value could not be inspected safely', path);
  }

  if (array) {
    // Array ORDER is data. It is never sorted — sorting an event sequence would
    // silently change what the figure says.
    const keys = safeOwnKeys(value as object, path);
    const lengthDescriptor = safeDescriptor(value as object, 'length', path);
    const length = lengthDescriptor.value;
    if (!Number.isSafeInteger(length) || length < 0) {
      throw new CanonicalizationError('array length is outside the canonical JSON domain', path);
    }
    const indexKeys: string[] = [];
    for (const key of keys) {
      if (typeof key === 'symbol') {
        throw new CanonicalizationError('symbol-keyed array members are outside the JSON domain', path);
      }
      if (key === 'length') continue;
      const index = Number(key);
      if (!/^(0|[1-9][0-9]*)$/u.test(key) || !Number.isSafeInteger(index) || index >= length) {
        throw new CanonicalizationError('named array members are outside the JSON domain', path);
      }
      const descriptor = safeDescriptor(value as object, key, childPath(path, key));
      if (!descriptor.enumerable || !Object.prototype.hasOwnProperty.call(descriptor, 'value')) {
        throw new CanonicalizationError('array accessors and hidden members are outside the JSON domain', path);
      }
      indexKeys.push(key);
    }
    if (indexKeys.length !== length) {
      throw new CanonicalizationError('sparse arrays are outside the canonical JSON domain', path);
    }
    indexKeys.sort((left, right) => Number(left) - Number(right));
    const parts: string[] = [];
    for (const key of indexKeys) {
      const at = childPath(path, key);
      const descriptor = safeDescriptor(value as object, key, at);
      parts.push(serialize(descriptor.value, at, depth + 1));
    }
    return `[${parts.join(',')}]`;
  }

  const record = value as Record<string, unknown>;
  let prototype: object | null;
  try {
    prototype = Object.getPrototypeOf(record);
  } catch {
    throw new CanonicalizationError('the object prototype could not be inspected safely', path);
  }
  if (prototype !== Object.prototype && prototype !== null) {
    throw new CanonicalizationError(
      'only plain objects can be canonicalized; a class instance has no canonical JSON form',
      path,
    );
  }

  // UTF-16 code-unit ordering, per RFC 8785 §3.2.3. Default sort() is exactly this.
  const ownKeys = safeOwnKeys(record, path);
  const keys: string[] = [];
  for (const key of ownKeys) {
    if (typeof key === 'symbol') {
      throw new CanonicalizationError('symbol-keyed members are outside the JSON domain', path);
    }
    const descriptor = safeDescriptor(record, key, childPath(path, key));
    if (!descriptor.enumerable || !Object.prototype.hasOwnProperty.call(descriptor, 'value')) {
      throw new CanonicalizationError('accessors and hidden members are outside the JSON domain', path);
    }
    keys.push(key);
  }
  keys.sort();

  const parts: string[] = [];
  for (const key of keys) {
    assertWellFormed(key, path);
    const at = childPath(path, key);
    const child = safeDescriptor(record, key, at).value;
    // `undefined` is not a JSON value. JSON.stringify silently DROPS such a
    // member, which would let two semantically different objects canonicalize to
    // the same bytes. Rejecting is the only safe answer.
    if (child === undefined) {
      throw new CanonicalizationError(
        `member ${JSON.stringify(key)} is undefined; undefined is not a JSON value`,
        at,
      );
    }
    parts.push(`${JSON.stringify(key)}:${serialize(child, at, depth + 1)}`);
  }
  return `{${parts.join(',')}}`;
}

/** Canonicalize a JSON-compatible value to its RFC 8785 byte sequence, as a string. */
export function canonicalize(value: unknown): string {
  return serialize(value, '', 0);
}

const UTF8 = new TextEncoder();

/** The canonical UTF-8 bytes. This is what actually gets hashed. */
export function canonicalBytes(value: unknown): Uint8Array {
  return UTF8.encode(canonicalize(value));
}

/**
 * SHA-256 over the canonical bytes of a value: `sha256:<64 hex>`.
 *
 * Two implementations that agree here agree on identity. That is the whole point.
 */
export function canonicalDigest(value: unknown): string {
  return sha256Digest(canonicalize(value));
}

/**
 * Digest an object with one top-level member excluded.
 *
 * An artifact carries its own digest, so that field cannot be part of what is
 * hashed — a self-referential hash has no fixed point. This makes the exclusion
 * explicit and testable rather than an implicit delete somewhere in the builder.
 */
export function canonicalDigestExcluding(
  value: Record<string, unknown>,
  excludeKey: string,
): string {
  const copy: Record<string, unknown> = Object.create(null);
  for (const key of safeOwnKeys(value, '')) {
    if (typeof key === 'symbol') {
      throw new CanonicalizationError('symbol-keyed members are outside the JSON domain', '');
    }
    const descriptor = safeDescriptor(value, key, childPath('', key));
    if (!descriptor.enumerable || !Object.prototype.hasOwnProperty.call(descriptor, 'value')) {
      throw new CanonicalizationError('accessors and hidden members are outside the JSON domain', '');
    }
    if (key !== excludeKey) copy[key] = descriptor.value;
  }
  return canonicalDigest(copy);
}

File: python/src/cortexel/canonicalize.py
"""RFC 8785 JSON Canonicalization Scheme — the independent Python implementation.

This is the second reader. If it and the TypeScript canonicalizer disagree on one byte,
every cross-language digest claim is false — so this is implemented from the scheme's
rules, not by reaching for ``json.dumps(sort_keys=True)``, which differs from JCS in
number formatting, key ordering, and escaping.

The two subtle points, both handled explicitly:

* Numbers use the ECMAScript ``Number::toString`` algorithm, NOT Python's ``repr``.
  Python prints ``100.0`` where JCS requires ``100``, and ``1e-06`` where JCS requires
  ``0.000001``. ``_js_number`` ports the ECMAScript formatting from the shortest
  round-tripping digits.
* ``-0.0`` canonicalizes to ``0`` (``-0.0 == 0.0`` in Python, so this falls out of the
  zero check — but only because it is written deliberately, not by accident).

Object members sort by UTF-16 code unit (RFC 8785 section 3.2.3). Python's default sort
is by code POINT, which differs for characters outside the Basic Multilingual Plane
(astral characters such as emoji). ``_utf16_key`` re-encodes to UTF-16 code units so the
ordering matches JavaScript exactly.
"""

from __future__ import annotations

import hashlib
import math
from typing import Any

_MAX_SAFE_INTEGER = (1 << 53) - 1


class CanonicalizationError(ValueError):
    """A value outside the JCS domain has no canonical form."""


def _js_number(value: float) -> str:
    """Format a finite number exactly as ECMAScript ``Number::toString`` would."""
    if math.isnan(value) or math.isinf(value):
        raise CanonicalizationError("non-finite numbers are outside the JCS domain")

    # -0.0 == 0.0, so both map here; JCS emits "0".
    if value == 0:
        return "0"

    negative = value < 0
    magnitude = -value if negative else value

    # Shortest round-tripping decimal digits and the decimal point position `n`, such that
    # value = s * 10**(n - k) with s a k-digit integer (10**(k-1) <= s < 10**k).
    digits, n = _shortest_digits(magnitude)
    k = len(digits)

    if k <= n <= 21:
        body = digits + "0" * (n - k)
    elif 0 < n <= 21:
        body = digits[:n] + "." + digits[n:]
    elif -6 < n <= 0:
        body = "0." + "0" * (-n) + digits
    else:
        # Exponential form: one digit, optional fraction, "e", signed exponent (n - 1).
        exponent = n - 1
        mantissa = digits[0] if k == 1 else digits[0] + "." + digits[1:]
        sign = "+" if exponent >= 0 else "-"
        body = f"{mantissa}e{sign}{abs(exponent)}"

    return "-" + body if negative else body


def _shortest_digits(magnitude: float) -> tuple[str, int]:
    """Return (significant-digit string without trailing zeros, decimal position n).

    ``repr(float)`` yields the shortest decimal that round-trips (Python 3), the same
    shortest set JavaScript uses; only the *formatting* differs, and that is what this
    function normalizes away.
    """
    text = repr(magnitude)

    mantissa, _, exp_part = text.partition("e")
    exponent = int(exp_part) if exp_part else 0

    int_part, _, frac_part = mantissa.partition(".")
    all_digits = int_part + frac_part
    # `n` counts digits before the decimal point in the un-normalized mantissa, then shifts
    # by the parsed exponent.
    point = len(int_part) + exponent

    # Strip leading zeros (adjusting the point) and trailing zeros.
    stripped = all_digits.lstrip("0")
    point -= len(all_digits) - len(stripped)
    stripped = stripped.rstrip("0")
    if stripped == "":
        stripped = "0"

    return stripped, point


def _utf16_key(key: str) -> list[int]:
    """A UTF-16 code-unit sort key, matching JavaScript's string comparison."""
    return list(key.encode("utf-16-be"))


def _assert_well_formed(text: str) -> None:
    # A lone surrogate has no UTF-8 encoding, so no canonical byte sequence.
    try:
        text.encode("utf-8")
    except UnicodeEncodeError as exc:
        raise CanonicalizationError("string is not well-formed Unicode") from exc


def _escape_string(text: str) -> str:
    _assert_well_formed(text)
    out = ['"']
    for ch in text:
        code = ord(ch)
        if ch == '"':
            out.append('\\"')
        elif ch == "\\":
            out.append("\\\\")
        elif ch == "\b":
            out.append("\\b")
        elif ch == "\f":
            out.append("\\f")
        elif ch == "\n":
            out.append("\\n")
        elif ch == "\r":
            out.append("\\r")
        elif ch == "\t":
            out.append("\\t")
        elif code < 0x20:
            out.append(f"\\u{code:04x}")
        else:
            out.append(ch)
    out.append('"')
    return "".join(out)


def canonicalize(value: Any, _depth: int = 0) -> str:
    """Canonicalize a JSON-compatible value to its RFC 8785 string form."""
    if _depth > 128:
        raise CanonicalizationError("value nests deeper than the canonicalizer permits")

    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, bool):  # pragma: no cover - covered by the two lines above
        return "true" if value else "false"
    if isinstance(value, int):
        # Python integers are arbitrary precision, while the JCS data model is IEEE-754.
        # Values outside the I-JSON safe range must be represented as strings or as an
        # explicitly approximate float; silently rounding them would create digest aliases.
        if abs(value) > _MAX_SAFE_INTEGER:
            raise CanonicalizationError("integer is outside the interoperable exact range")
        return str(value)
    if isinstance(value, float):
        return _js_number(value)
    if isinstance(value, str):
        return _escape_string(value)
    if isinstance(value, (list, tuple)):
        # Array order is data; never sorted.
        return "[" + ",".join(canonicalize(item, _depth + 1) for item in value) + "]"
    if isinstance(value, dict):
        keys = sorted(value.keys(), key=_utf16_key)
        parts = []
        for key in keys:
            if not isinstance(key, str):
                raise CanonicalizationError("object keys must be strings")
            child = value[key]
            parts.append(f"{_escape_string(key)}:{canonicalize(child, _depth + 1)}")
        return "{" + ",".join(parts) + "}"

    raise CanonicalizationError(f"values of type {type(value).__name__} are outside the JCS domain")


def canonical_digest(value: Any) -> str:
    """SHA-256 over the canonical bytes: ``sha256:<64 hex>``. Uses the standard library
    hashlib — an independent, well-audited SHA-256, not a reimplementation."""
    canonical = canonicalize(value).encode("utf-8")
    return "sha256:" + hashlib.sha256(canonical).hexdigest()


def canonical_digest_excluding(value: dict, exclude_key: str) -> str:
    """Digest an object with one top-level member removed (for a self-referential digest)."""
    copy = {k: v for k, v in value.items() if k != exclude_key}
    return canonical_digest(copy)
